eval.c

```diff
@@ -628,112 +639,75 @@ DO_OP_STORE_GLOBAL: {
 	DISPATCH();
 }
 
-DO_OP_NEW_FUNCTION: {
-	size_t name, arity, address, nopt, i, stack_required, nlocals;
-	struct cb_function *func;
-	struct cb_value func_val;
-
-	name = READ_SIZE_T();
-	arity = READ_SIZE_T();
-	address = READ_SIZE_T();
-	nlocals = READ_SIZE_T();
-	stack_required = READ_SIZE_T();
-	nopt = READ_SIZE_T();
-
-	func = cb_function_new();
-	func->type = CB_FUNCTION_USER;
-	func->name = name;
-	func->arity = arity - nopt;
-	func->value.as_user = (struct cb_user_function) {
-		.address = address,
-		.upvalues = NULL,
-		.upvalues_len = 0,
-		.upvalues_size = 0,
-		.module_id = frame->module
-			? cb_modspec_id(frame->module->spec)
-			: (size_t) -1,
-		.nlocals = nlocals,
-		.stack_required = stack_required,
-		.optargs = (struct cb_function_optargs) {
-			.count = nopt,
-		},
-	};
-
-	for (i = 0; i < nopt; i += 1)
-		func->value.as_user.optargs.addrs[i] = READ_SIZE_T();
-
-	func_val.type = CB_VALUE_FUNCTION;
-	func_val.val.as_function = func;
-
-	PUSH(func_val);
-	DISPATCH();
-}
-
 DO_OP_BIND_LOCAL: {
 	int i;
-	struct cb_value *func;
+	struct cb_value func;
 	struct cb_upvalue *uv;
+	size_t dest_idx = READ_SIZE_T();
 	size_t idx = LOCAL_IDX(READ_SIZE_T());
 
-	func = &TOP();
-	if (func->type != CB_VALUE_FUNCTION
-			|| func->val.as_function->type != CB_FUNCTION_USER)
+	func = POP();
+	if (func.type != CB_VALUE_FUNCTION
+			|| func.val.as_function->type != CB_FUNCTION_USER)
 		ERROR("Can only bind upvalues to user functions");
 
 	uv = NULL;
-	for (i = state->upvalues_idx - 1; i >= 0; i -= 1) {
-		if (state->upvalues[i] == NULL
-				|| !state->upvalues[i]->is_open
-				|| state->upvalues[i]->v.idx < frame->bp)
+	for (i = cb_vm_state.upvalues_idx - 1; i >= 0; i -= 1) {
+		if (cb_vm_state.upvalues[i] == NULL
+				|| cb_vm_state.upvalues[i]->call_depth < 0
+				|| (size_t) cb_vm_state.upvalues[i]->call_depth
+					< cb_vm_state.call_depth)
 			break;
-		if (state->upvalues[i]->v.idx == idx) {
-			uv = state->upvalues[i];
+		if (cb_vm_state.upvalues[i]->v.idx == idx) {
```

Comment:
i dunno if matching on idx here is sufficient to uniquely identify an upvalue... if call_depth is the depth of the frame where the upvalue was originally declared, then maybe this can look directly at the function for that frame to search for it

Reply:
actually this only needs to look at the upvalues for the current function i think?